### PR TITLE
fsautocomplete: init at 0.58.2

### PR DESCRIPTION
**Pull request by gbtb**
###### Description of changes

I've packaged a dotnet tool for F# Language Server (fsautocomplete) straight from its nuget package.
Right now it's crudely built on top of buildDotnetModule, and I'm open to suggestions for how maybe extract some of the code into reusable blocks.
Hoping to get a review from nixpkgs's dotnet masterminds @IvarWithoutBones @mdarocha :smiley: 
Related to an issue https://github.com/NixOS/nixpkgs/issues/178386

###### Things done

29 packages built:
ArchiSteamFarm BeatSaberModManager alttpr-opentracker baget boogie btcpayserver depotdownloader discordchatexporter-cli eventstore formula fsautocomplete galaxy-buds-client inklecate jackett jellyfin msbuild nbxplorer netcoredbg omnisharp-roslyn opentabletdriver osu-lazer pinta ps3-disc-dumper python-language-server roslyn ryujinx scarab wasabibackend xivlauncher

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix [hunk 1]
```diff
@@ -54,11 +55,12 @@ in
     } ./dotnet-install-hook.sh) { };
 
   dotnetFixupHook = callPackage ({ }:
+    let dep = if isDotnetTool then dotnet-sdk else dotnet-runtime; in
```
Comment (mdarocha) [inline pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix:58]: Why is this needed?

File: pkgs/development/tools/fsautocomplete/default.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
+{ buildDotnetModule, mkNugetDeps, emptyFile, lib, dotnet-sdk }:
```
Comment (mdarocha) [inline pkgs/development/tools/fsautocomplete/default.nix:1]: Overall looks ok, however I would really prefer if the package was built from source instead of downloaded from nuget

File: pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix (hunk 1 above)
Comment (gbtb) [reply pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix:58]: All dotnet tools need full sdk in order to work. I guess it's logical because usually tools are used at build-time.

File: pkgs/development/tools/fsautocomplete/default.nix (hunk 2 above)
Comment (gbtb) [reply pkgs/development/tools/fsautocomplete/default.nix:1]: I've tried to do it, but it turned to be quite difficult. Unhelpfully, this package itself uses another tool - paket - for managing its nuget dependencies, so it's a bit a chicken and egg problem. I guess we can use this "binary" install approach as a stage-1 and later develop another method for building tools from sources.

File: pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix (hunk 1 above)
Comment (mdarocha) [reply pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix:58]: Hmm, can you maybe make the flag a bit more generic? Like renaming it to 'useSdkAsRuntime', and of course documenting it properly in dotnet.section.md

File: pkgs/development/tools/fsautocomplete/default.nix (hunk 2 above)
Comment (mdarocha) [reply pkgs/development/tools/fsautocomplete/default.nix:1]: Fine, I personally had a couple problems with paket and Nix. We'll have to develop a solution to that, and then the package can be migrated to source build

**Review by mdarocha — APPROVED** [review]

File: pkgs/development/tools/fsautocomplete/default.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+let
+  fsautocomplete = buildDotnetModule rec {
+    pname = "fsautocomplete";
+    version = "0.58.2";
+    src = emptyFile;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/fsautocomplete/default.nix:6]: Please use dontUnpack instead

File: pkgs/development/tools/fsautocomplete/default.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    inherit dotnet-sdk;
+    nugetDeps = mkNugetDeps {
+      name = pname;
+      nugetDeps = { fetchNuGet }: [ (fetchNuGet { inherit pname version; sha256 = "sha256-xZFTxdD4ma5IlPZNqOSwJv/1ixicm21FY+WVS574QiI="; }) ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/tools/fsautocomplete/default.nix:11]:
```suggestion
      nugetDeps = { fetchNuGet }: [
        (fetchNuGet { inherit pname version; sha256 = "sha256-xZFTxdD4ma5IlPZNqOSwJv/1ixicm21FY+WVS574QiI="; })
      ];
```

File: pkgs/development/tools/fsautocomplete/default.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    };
+    dontUnpack = true;
+    dontInstall = true;
+    useSdkAsRuntime = true;
+    configurePhase = ''
+         # Generate a NuGet.config file to make sure everything,
+         # including things like <Sdk /> dependencies, is restored from the proper source
```
Comment (SuperSandro2000) [inline pkgs/development/tools/fsautocomplete/default.nix:18]:
```suggestion
    # Generate a NuGet.config file to make sure everything,
    # including things like <Sdk /> dependencies, is restored from the proper source
```

File: pkgs/development/tools/fsautocomplete/default.nix [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    '';
+
+    meta = with lib; {
+      description = "The FsAutoComplete project (FSAC) provides a backend service for rich editing or intellisense features for editors.";
+      homepage = "https://github.com/fsharp/FsAutoComplete";
+      changelog = "https://github.com/fsharp/FsAutoComplete/releases/tag/v${version}";
+      license = licenses.asl20;
+      platforms = platforms.linux;
+      maintainers = with maintainers; [ gbtb ];
+      mainProgram = pname;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/fsautocomplete/default.nix:49]:
```suggestion
```
that is the default

File: pkgs/development/tools/fsautocomplete/default.nix (hunk 3 above)
Comment (gbtb) [reply pkgs/development/tools/fsautocomplete/default.nix]: Thanks. Fixed all issues you've mentioned.

**Review by ryantm — APPROVED** [review]
Looks good to me and builds.

File: pkgs/development/tools/fsautocomplete/default.nix [hunk 7]
```diff
@@ -0,0 +1,52 @@
+{ buildDotnetModule, mkNugetDeps, emptyFile, lib, dotnet-sdk }:
+let
```
Comment (mdarocha) [inline pkgs/development/tools/fsautocomplete/default.nix:2]: this doesnt seem to need to be in a `let in` binding, instead have the `buildDotnetModule` at the root of the file

**Review by mdarocha — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/fsautocomplete/default.nix (hunk 7 above)
Comment (ratsclub) [reply pkgs/development/tools/fsautocomplete/default.nix:2]: @mdarocha this seems to be needed for this line: `<add key="nugetSource" value="${fsautocomplete.passthru.nuget-source}/lib" />`

Comment (mdarocha) [reply pkgs/development/tools/fsautocomplete/default.nix:2]: Oh, didn't notice that

**Review by mdarocha — CHANGES_REQUESTED** [review]
As I posted in https://github.com/NixOS/nixpkgs/issues/216285, I noticed a lot of duplication across similar PRs.

Try to create a `buildDotnetTool` abstraction, that uses `buildDotnetModule` and sets all the common stuff (ie. `configurePhase`, `buildPhase`, `executables`, etc.) in one place. You can place it in `pkgs/build-support/dotnet/bulid-dotnet-tool` for example.

I wouldn't want to merge these dotnet tool PRs in the current state, due to the mentioned duplication, it would be hard to maintain.

I would also prefer it if the package was built from source, have you tried it?

File: pkgs/development/tools/fsautocomplete/default.nix (hunk 7 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/fsautocomplete/default.nix:2]: see the finalAttrs here https://github.com/NixOS/nixpkgs/blob/49966f2c7139521b1c84af4c76c72725e1eafd2e/doc/stdenv/meta.chapter.md#nixos-tests to avoid this

**Comment by SuperSandro2000** [comment]
closes by #235041
